c_tests/tests/compile_trace_not_entry_bb.c

```diff
@@ -0,0 +1,34 @@
+// Compiler:
+// Run-time:
+
+// Check that trace compilation works in the non-entry block.
+//
+// Since LLVM allocas typically appear in the entry block of a function, we
+// will miss the allocas if tracing starts in a later block.
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <yk_testing.h>
+
+int main(int argc, char **argv) {
+  // Causes the traced block to NOT be the entry block.
+  if (argc == -1)
+    abort();
+
+  int res;
+  void *tt = __yktrace_start_tracing(HW_TRACING);
+  // Causes both a load and a store to things defined outside the trace.
+  res = 1 + argc;
```

Comment:
This test needs a FIXME as, when optimised, this will quite probably turn into:

```
start_tracing();
stop_tracing();
assert(argc == 1);
```

Reply:
We can't go on like this. Do we have a solution to stop things floating around?

I have a branch which checks all optimisation levels, but i'd prefer a proper fix.

Reply:
is the asm volatile trick acceptable?

Reply:
> Do we have a solution to stop things floating around?

There is no guaranteed solution. *But* you're *much* less likely to have problems if things are structured along the lines of:

```
fn bench() {
  start_tracing();
  f();
  stop_tracing();
}

#[inline(never)]
fn f() {
  // the benchmark
}
```

perhaps using `asm` tricks and so on. Put another way: having benchmark code in the same function as the timing loop all but guarantees that the optimising compiler will optimise things away.

Reply:
This is not going to work, because that's exactly what this test is supposed to check. If we put the benchmark into a function, then we will trace the allocas.

Reply:
I've recently noticed that LTO won't listen to requests to not inline.

> There is no guaranteed solution

I think we can find platform specific solutions, like using a `MFENCE` and an compiler barrier (i.e. mfence in a volatile asm block).

![image](https://user-images.githubusercontent.com/604955/119809370-91006b00-bed4-11eb-89fd-b056a1512f18.png)

Reply:
Good point. Perhaps this PR is solving the wrong problem in a sense? *Perhaps* what we should actually do is drop the `start_tracing`/`stop_tracing` model and move to a `trace(f)` model which means "give me back a trace of function `f`". That way we don't have to worry about this alloca problem at all, and we can also (probably) stop the compiler optimising `f` into `trace`.

Reply:
What are the implications for the VM author?

Reply:
Ouch, yes, sometimes my suggestion would make their life miserable (though sometimes it would work). Scratch that (at least until I've thought more about it).

MFENCE etc. stops *some* optimisations (side effecting things) but not all. You're also right that nothing guarantees to stop inlining, LTO or otherwise. There are no great solutions here. That said, `start_tracing` and `stop_tracing` can easily have compiler barriers added such that side effecting optimisations won't move. However, notice that this test does not have side effects, so compiler barriers won't help.

Reply:
For this PR, let's just add a FIXME to this test.